### PR TITLE
[HttpClient] Allow setting a connect timeout in `CurlHttpClient`

### PR DESCRIPTION
**Pull request by HypeMC**
| Q             | A
| ------------- | ---
| Branch?       | 7.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | -
| License       | MIT

It'd be useful to be able to set the connect time in the Curl client.

The project I'm working on uses the client to send purge requests to our Varnish servers. Usually, the requests take less than half a second to complete, but occasionally, due to various reasons, a certain request can take 2-3 seconds. Because of that, the `max_duration` is set to 5 seconds.

Problems start occurring when one or more of the Varnish servers are down, whether for maintenance or other reasons. When that happens, all requests wait up to 5 seconds. Since the number of requests is significant, the message queue starts filling up as the workers slow down and can't keep up with the number of queued URLs.

The way we would fix this before switching to the HttpClient component was by setting the `CURLOPT_CONNECTTIMEOUT_MS` to 100ms. If the request was not able to connect to the server within that time, we'd assume it's down. Otherwise, the request would be performed as expected.

### CODE REVIEW

File: src/Symfony/Component/HttpClient/Tests/CurlHttpClientTest.php [hunk 1]
```diff
@@ -120,4 +121,24 @@ public function testOverridingInternalAttributesUsingCurlOptions()
...
+
+    /**
+     * @testWith [1]
+     *           [5]
+     */
+    public function testOverridingConnectTimeoutUsingCurlOptions(int $connectTimeout)
+    {
+        $httpClient = $this->getHttpClient(__FUNCTION__);
+        $this->expectException(TransportException::class);
+        $this->expectExceptionMessageMatches(\sprintf('~Connection (timeout|timed out) after %d (ms|milliseconds) for "http\://10\.255\.255\.1/"\.~', $connectTimeout));
```
Comment (HypeMC) [inline src/Symfony/Component/HttpClient/Tests/CurlHttpClientTest.php:133]: The message I got locally is slightly different than the one on the CI, I'm guessing it depends on the version of Curl.

**Comment by nicolas-grekas** [comment]
The timeout should be handled on your side by using the second argument of the `$client->stream()` method. That's the abstract and portable way to manage timeouts.

**Comment by HypeMC** [comment]
@nicolas-grekas The timeout option doesn't seem to give me the expected result. When used with a request that takes longer, it times out, which is not what I want. However, when used with a non-routable IP, it doesn't time out as expected, in fact, it takes much longer. For example:

```php
$curlClient = new \Symfony\Component\HttpClient\CurlHttpClient([
    'max_duration' => 5,
]);

$response = $curlClient->request('GET', 'http://localhost:8090');
// $response = $curlClient->request('GET', 'http://10.255.255.255');

foreach ($curlClient->stream($response, 0.01) as $key => $value) {
    echo $value->getContent();
}
```

where `localhost:8090` is `php -S localhost:8090 index.php` with:

```php
<?php

sleep(5);
echo time();
```

On the other hand, the `CURLOPT_CONNECTTIMEOUT` option gives me exactly what I need. The request to `localhost:8090` finishes successfully, while a request to a non-routable IP terminates within 10ms:

```php
$curlClient = new \Symfony\Component\HttpClient\CurlHttpClient([
    'max_duration' => 5,
    'extra' => [
        'curl' => [
            \CURLOPT_CONNECTTIMEOUT_MS => 10,
        ]
    ],
]);

echo $curlClient->request('GET', 'http://localhost:8090')->getContent(), PHP_EOL;
// echo $curlClient->request('GET', 'http://10.255.255.255')->getContent(), PHP_EOL;
```

**Comment by nicolas-grekas** [comment]
~Locally I have the exact same behavior with both requests. Note that you must comment at least one line to be able to compare. If you uncomment both lines, then yes, the 5s timeout will apply thanks to the destructor of the 1st line.~

**Comment by nicolas-grekas** [comment]
Ah nvm, I get what you mean now.

**Comment by nicolas-grekas** [comment]
The following works.
It relies on the `connect_time` info, which is not in the contracts but is still provided by all implementations.

```php
//$response = $client->request('GET', 'http://localhost:8090');
$response = $client->request('GET', 'http://10.255.255.255');

foreach ($client->stream($response, 0.1) as $chunk) {
    if ($chunk->isTimeout() && !$response->getInfo('connect_time')) {
        $response->cancel();

        throw new \Exception('Connect timeout');
    }

    break;
}

dump($response->getContent());
```

**Comment by HypeMC** [comment]
@nicolas-grekas Wow, nice! As far as I can tell, this works for every case I could think of. Thank you for your help, I'm closing this one now.

**Comment by OskarStark** [comment]
@HypeMC do you think this is important and should be added to the documentation?

**Comment by HypeMC** [comment]
@OskarStark Personally, I'd say yes, it would be good to have this documented, as it's not obvious at first.

**Comment by nicolas-grekas** [comment]
There are hundred of such snippets we could add to the doc. Dunno how that could fit - this can be scary also to have so many detailed use cases.

**Comment by HypeMC** [comment]
Fair point. This is a very specific use case, and probably not many people need something like this.

**Comment by voodooism** [comment]
@nicolas-grekas
I've encountered a similar issue. In my case, I'm using a third-party library that accepts an HttpClient as a dependency. It would be extremely useful to have an option to configure `CURLOPT_CONNECTTIMEOUT` in the config. I don’t want to wait longer than the `CURLOPT_CONNECTTIMEOUT` value for DNS + TCP + TLS, but currently there is no way to configure this.

Why can’t we pass `CURLOPT_CONNECTTIMEOUT` as an extra.curl option?

**Comment by OskarStark** [comment]
Can you please open a new issue @voodooism ? Thanks

**Pull request closed without merging**